### PR TITLE
IPA-684 Start job batch will put to kafka queue.

### PR DESCRIPTION
**Pull request by shirleyqiantingliu**
Spreading evens support both kafka and rabbitMQ events.

# Problem / Feature
Start job batch might take long time, it should be put to kafka event first, and then really starting the job batch.

### CODE REVIEW

File: src/api/emitter/index.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const KafkaClient = require('@opuscapita/kafka-client');
+const EventClient = require('@opuscapita/event-client');
+const { JOB_BATCH_START_EVENT } = require('../../utils/constants');
+const { startJobBatchHandler } = require('./startJobBatchHandler');
+
+class Emitter {
```
Comment (conckja1) [inline src/api/emitter/index.js]: Is Emitter a good name since we are also subscribing... We should maybe use some generalization

Comment (shirleyqiantingliu) [reply src/api/emitter/index.js]: What would be better?

File: package.json [hunk 2]
```diff
@@ -35,6 +37,7 @@
     "mocha": "7.0.1",
     "nyc": "15.0.0",
     "path": "0.12.7",
+    "proxyquire": "^2.1.3",
```
Comment (conckja1) [inline package.json]: Do we need this in dependencies or should it be in devDependencies?

Comment (shirleyqiantingliu) [reply package.json]: right

Comment (conckja1) [reply package.json]: I am not 100% sure, but packages that are used only in unit-tests should go to devDependencies? At least in Invoice it seems to be done that way

File: src/api/emitter/index.js (hunk 1 above)
Comment (conckja1) [reply src/api/emitter/index.js]: Can't figure out anything really better. Maybe Messenger. just keep it as it is.

File: src/api/jobBatch/index.js [hunk 3]
```diff
@@ -92,6 +96,80 @@ const finishJobBatch = async (storage, jobId) => {
...
+  }
+  await storage.updateJobBatch(jobBatchId, {
+    progressStatus: PROGRESS_STATUS.IN_PROGRESS
+  });
+  await executeActions(storage, jobs, jobBatch.actions, logger);
+}
+
 class JobBatchApi {
 
   static getInstance(storage) {
```
Comment (conckja1) [inline src/api/jobBatch/index.js]: Just a question: what is the reason of making this class as singleton?

File: src/api/index.js [hunk 4]
```diff
@@ -1,16 +1,39 @@
...
-  const storage = getStorage(db);
-  JobBatchApi.getInstance(storage);
-  JobApi.getInstance(storage);
+const init = async (app, db, conf) => {
+  try {
+    const storage = getStorage(db);
+    const emitter = new Emitter(storage, conf && conf.serviceName);
+    await emitter.init();
+    JobBatchApi.getInstance(storage);
+    JobApi.getInstance(storage);
```
Comment (conckja1) [inline src/api/index.js]: Also here, why is `class JobApi` made as singleton? Usually the chief point of doing so is to share some resource, like a DB or rabbitMQ server connection. But this class doesn't seem to have any.

File: src/api/emitter/index.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  static getInstance(storage) {
+    if (Emitter.INSTANCE) {
+      return Emitter.INSTANCE;
+    }
+    return new Emitter(storage);
+  }
+
+  constructor(storage, serviceName) {
```
Comment (conckja1) [inline src/api/emitter/index.js]: I dont have the whole picture, where is the parameter `serviceName` passed?

File: src/api/index.js (hunk 4 above)
Comment (shirleyqiantingliu) [reply src/api/index.js]: Yes, the only reason here is to share the storage, which is from db. It makes the logic more complicated. I am thinking to make it more simple somehow.

File: src/api/emitter/index.js (hunk 5 above)
Comment (shirleyqiantingliu) [reply src/api/emitter/index.js]: It is passed when it is initialized from the config. if you check api/index.js

**Pull request merged**